lib/metasploit/framework/data_service/proxy/host_data_proxy.rb

```diff
@@ -97,7 +97,8 @@ def delete_host_tag(opts)
         return data_service.delete_host_tag(opts)
       end
     else
-      nil
+      data_service = self.get_data_service
+      return data_service.delete_host_tag(opts)
     end
```

Comment:
Is this now just:

```
  def delete_host_tag(opts)
      self.data_service_operation do |data_service|
        return data_service.delete_host_tag(opts)
      end
  end
```

Reply:
Not really cause if the data service doesn't exist, aka `@current_data_service` is empty, then we will raise an exception when we call `get_data_service` in `data_service/proxy/core.rb`. With the `data_service_operation` call we will return `nil` or pass the data to the passed in block.

Unless I'm missing something here?

Reply:
I believe the preference - if you're not connected to a database then the database operations should be noop, instead of raising an exception. Unless I am also missing something here 😄

Reply:
Not sure what we would prefer here given I don't have enough context as to how this should be implemented. If someone has a better idea would appreciate a confirmation on this.

Reply:
Alan is right, it looks these updates just ignore the result of `#add_host_id_to_opts` and move on to obtaining the data service before calling `#delete_host_tag`. [`#data_service_operation`](https://github.com/rapid7/metasploit-framework/blob/9ef9f9889407783d259314b3bb5d0912df5502bc/lib/metasploit/framework/data_service/proxy/core.rb#L155-L165) just returns obtains the data service and passes it to the specified block.

I would have expected an exception to be raised, but that doesn't seem to have happened. Either way, the break point I've set is not triggered in this function when the data service is disconnected.

```
msf6 auxiliary(admin/dcerpc/icpr_cert) > hosts -T mytag
msf6 auxiliary(admin/dcerpc/icpr_cert) > db_status
[*] Connected to metasploit. Connection type: postgresql.
msf6 auxiliary(admin/dcerpc/icpr_cert) > db_disconnect
Successfully disconnected from the data service: local_db_service.
msf6 auxiliary(admin/dcerpc/icpr_cert) > hosts -T mytag
[-] Database not connected
msf6 auxiliary(admin/dcerpc/icpr_cert) >
```

@gwillcox-r7 Is there a case where `#add_host_id_to_opts` will return false and it's still necessary to call `#delete_host_tag`? If so then you should update this with the [original suggestion](https://github.com/rapid7/metasploit-framework/pull/17576/files#r1224577732) and if not, it should be reverted back to `nil`.

Reply:
I'm going to change this back to `nil`. Further inspection of the use cases of this shows that the same pattern is used in at least 2-3 other places we aren't changing so this would lead to inconsistency if we implemented it in the current fashion. Furthermore I don't see any cases from the 4 use cases SourceGraph was able to find where one would want to delete an tag if the ID of the host could not be found.

Thanks for calling this out @adfoster-r7 and for the clarification @smcintyre-r7!

Reply:
Should be fixed now with last push, let me know if this needs anything else.